### PR TITLE
Update program attributes

### PR DESCRIPTION
**Pull request by ahsan-ul-haq**
ECOM-2752

* Update Program API to accept patch requests.
* Write Program test to check update program attributes using patch.

@awais786 @zubair-arbi @tasawernawaz @afeef please review.

### CODE REVIEW

File: programs/apps/api/v1/tests/test_views.py [hunk 1]
```diff
@@ -155,6 +160,41 @@ def test_create(self):
             }
         )
 
+    def test_patch(self):
```
Comment (awais786) [inline programs/apps/api/v1/tests/test_views.py]: use ddt for different keys.

File: programs/apps/api/v1/tests/test_views.py [hunk 2]
```diff
@@ -155,6 +160,41 @@ def test_create(self):
...
+        """
+        Ensure the API supports Patch request for Programs
+        """
+        program = ProgramFactory.create()
+        data = json.dumps({"subtitle": "dummy-subtitle"})
+        response = self._make_request(method='patch', program_id=program.id, data=data, admin=True)
+        self.assertEqual(response.status_code, 200)
+        data = json.loads(response.content)
+        self.assertEqual(data['subtitle'], "dummy-subtitle")
```
Comment (awais786) [inline programs/apps/api/v1/tests/test_views.py]: Also assert the whole response.content with some predefined dict.

File: programs/apps/api/v1/tests/test_views.py [hunk 3]
```diff
@@ -155,6 +160,41 @@ def test_create(self):
             }
         )
 
+    def test_patch(self):
+        """
+        Ensure the API supports Patch request for Programs
+        """
+        program = ProgramFactory.create()
+        data = json.dumps({"subtitle": "dummy-subtitle"})
```
Comment (zubair-arbi) [inline programs/apps/api/v1/tests/test_views.py]: nit: Please be consistent with the prefix, either user `dummy-` or `test-` for all strings

File: programs/apps/api/v1/tests/test_views.py [hunk 4]
```diff
@@ -155,6 +160,41 @@ def test_create(self):
...
+        response = self._make_request(method='patch', program_id=program.id, data=data, admin=True)
+        self.assertEqual(response.status_code, 200)
+        data = json.loads(response.content)
```
Comment (zubair-arbi) [inline programs/apps/api/v1/tests/test_views.py]: nit: rename `data` -> `response_data`

**Comment by zubair-arbi** [comment]
looks good :+1:

**Comment by ahsan-ul-haq** [comment]
@jimabramson please review.
FYI @rlucioni

File: programs/apps/api/v1/tests/test_views.py [hunk 5]
```diff
@@ -51,7 +51,12 @@ def _make_request(self, method='get', program_id=None, data=None, admin=False):
             url = reverse("api:v1:programs-detail", kwargs={'pk': program_id})
         else:
             url = reverse("api:v1:programs-list")
-        return getattr(self.client, method)(url, data=data, HTTP_AUTHORIZATION=auth)
+        if method == 'patch':
+            return getattr(self.client, method)(
+                url, data=data, HTTP_AUTHORIZATION=auth, content_type="application/json"
```
Comment (jimabramson) [inline programs/apps/api/v1/tests/test_views.py]: I think "application/json" is actually correct for all request methods; however, just had a conversation with some folks here about how we do PATCH.   The content_type in that case should be "application/merge-patch+json" for this case.  Take a look at the details here: https://openedx.atlassian.net/wiki/display/AC/edX+REST+API+Conventions#edXRESTAPIConventions-3.HTTPVerbs

In order to use that content type we need to add an extra bit to DRF to understand it.  There is an example in edx-platform [here](https://github.com/edx/edx-platform/blob/master/openedx/core/lib/api/parsers.py#L80-L84).  I suggest this should be added to a new module, programs/apps/api/parsers.py

File: programs/apps/api/v1/views.py [hunk 6]
```diff
@@ -32,6 +33,15 @@ class ProgramsViewSet(
         Only users with global administrative rights may create programs.  POST requests from non-
         admins will result in status 403.
 
+        # Update existing Program.
+        PATCH /api/v1/programs/{program_id}
+
+        If the request is successful, the HTTP status will be 200 and the response body will
+        contain a JSON-formatted representation of the newly-updated program field.
```
Comment (jimabramson) [inline programs/apps/api/v1/views.py]: minor correction: s/newly-updated program field/newly-updated program/

File: programs/apps/api/v1/views.py [hunk 7]
```diff
@@ -39,6 +49,7 @@ class ProgramsViewSet(
         * subtitle: A brief, descriptive subtitle for the Program.
         * category: The category / type of Program.  Right now the only value allowed is 'xseries'.
         * status: The lifecycle status of this Program.  Right now the only value allowed is 'unpublished'.
+        * marketing_slug: Slug used to generate links to the marketing site.
```
Comment (jimabramson) [inline programs/apps/api/v1/views.py]: thanks for fixing that

**Comment by jimabramson** [comment]
@ahsan-ul-haq looks solid.  Other than changes for merge-patch (see inline), looks good to go

File: programs/apps/api/v1/tests/test_views.py (hunk 5 above)
Comment (rlucioni) [inline programs/apps/api/v1/tests/test_views.py]: @jimabramson just so I'm clear, we can use DRF's default PATCH handling, but we need to add support for the merge patch media type?

Comment (ahsan-ul-haq) [inline programs/apps/api/v1/tests/test_views.py]: @jimabramson I guess that would be reasonable as per the conventions used in the edX. Thanks for sharing the resources.

**Comment by ahsan-ul-haq** [comment]
@jimabramson suggested changes have been addressed. It would be great if you could take a look again.

**Comment by jimabramson** [comment]
:+1: 
@rlucioni to your [question](https://github.com/edx/programs/pull/31#discussion_r44079944) - yes, correct.
@AlasdairSwan FYI - client API calls should use content type "application/merge-patch+json" when updating individual fields, based on the discussion/changes in this PR.

**Pull request merged**